### DISCUSSION TITLE
Self trained ONNX model

### DISCUSSION THREAD

**Discussion by NeilDale** [PUTvision/qgis-plugin-deepness#176] [General]
Hi,

I am a civil engineer who has some experience in GIS and very little programming experience and I came across your plug-in and I am trying to find a way to use a self trained model on it. I have trained a yolov7 model on +-4000 images of solar panels. I have exported it to ONNX but when I run the plugin in qgis I get the following error: "Error! Processing error: "Unhandled exception occured. See Python Console for detains"!

Details: Type Error - 'NonType' object is not subscriptable Last Traceback: File"---deepness\processing\models\detector.py", line 140, in get_number_of_output_channels
class_names = self.get_outputs_channel_names()[0]

could you please take a look and point me in the right direction to solve it please. 

Thank you

Neil

**Comment by bartoszptak** [comment]
Hi @NeilDale!
It looks like a problem with class names in the model's metadata. The best way to help you is to debug your model. Can you share it? You can send it to me by mail; the address is in my profile description.

**Reply by NeilDale** [reply]
Thank you so much.

I have sent the model.

**Reply by bartoszptak** [reply]
For some reason unknown to me, your model does not contain metadata after export. You can add it manually using [this](https://qgis-plugin-deepness.readthedocs.io/en/latest/creators/creators_add_metadata_to_model.html) guide.

In more detail, metadata allows you to define the behaviour and default parameters for the models. One such parameter is the `class_names` variable, which specifies what name the model outputs should use. I was sure that the YOLOv6-YOLOv9 model export provides such metadata, but I could be wrong. Deepness shall also automatically label the outputs `output_1` in case of an undefined `class_names` variable. We need to look into this.

**Comment by NeilDale** [comment]
Bartosz,

Thanks for the swift reply.

I will give it a try and I will let you know of the outcome, i am a novice
so it might take some time.

If you are available for a call some time, it would be great just to meet
you and have a chat. It's always nice to learn from experiences like yours.
I am available almost any time

Kind Regards

Neil

**Comment by NeilDale** [comment]
Hi,

Imanaged to follow the instructions and it works very well on pleiades neo
imagery that I have. as long as I select pixel size at 10cm.

Please feel free to try it out: solar-panel_detect.onnx
<https://1drv.ms/u/s!AqNZ3aqm1jXeiMFNrGQmfmuHjPLC7A?e=sSznbL>.

I will be happy to receive any feedback if you have any.

Thanks again for your time.

Kind Regards

**Comment by AlainFELER** [comment]
Hello, I tried your model on two small places where I know there is at least one solar panel, but it did not find them. Maybe it is because the local orthophoto pixel size is 20 cm and not 10. Maybe I can find a 10 cm cover, but is that the probably main problem ?

Also, could you explain in some detail how you produced the model ? I would like to make a detector model myself, for which I can also provide thousands of sample images, but I don't know how to proceed. I am sure it would be useful to many people.

**Reply by NeilDale** [reply]
Hi Alain,

Small disclaimer, I lucked out on getting this model to work and i'm far from an expert in the field.

I tested it on Pleiades neo (30cm imagery)and it could not detect any solar panels, but when i lied to the plugin and said that the pixels are 10cm, it worked. Maybe try smaller pixels values and see what happens.

Regards
